crates/p2p_protocol/protocol/p2p.capnp

```diff
@@ -0,0 +1,179 @@
+
+    msgQueryResponse @10 :MsgQuery;
+    # dht query response
+  }
+
+  # -- top-level Message Types -- #
+
+  struct MsgError {
+    errorCode @0 :ErrorCode;
+    # code indicating if error is well-known
+
+    errorText @1 :Text;
+    # text indicating details of error
+
+    enum ErrorCode {
+      unknown @0;
+      # default if error is not well-known, or if remote is using a newer proto
+    }
+  }
+
+  struct MsgHandshake {
+    fromStoreArc @0 :UInt32;
+    # the storage arc length of the source transport / agent
+
+    fromQueryArc @1 :UInt32;
+    # the query arc length of the source transport / agent
+  }
+
+  struct MsgGspArc {
+    # data for `msgGspArcRequest` and `msgGspArcResponse`
+
+    aspectConstraintLoc @0 :AspectConstraintLoc;
+    # constrain aspectHashList by loc arc
+
+    aspectConstraintTime @1 :AspectConstraintTime;
+    # constrain aspectHashList by publish timestamp
+
+    aspectConstraintCount @2 :AspectConstraintCount;
+    # constrain aspectHashList by local store count
+
+    aspectHashList @3 :List(AspectHashList);
+    # list of aspectHashes associated with entryAddresses
+    # that fall within all the above constraints.
+  }
+
+  struct MsgGspAspectDataRequest {
+    # request a list of aspect hashes
+
+    aspectHashList @0 :List(AspectHashList);
+    # the aspect hashes we are requesting
+  }
+
+  struct MsgGspAspectDataResponse {
+    # respond to an aspectDataRequest with aspect data
+
+    aspectHashDataList @0 :List(AspectHashDataList);
+    # the aspect data to respond with
+  }
+
+  struct MsgGspAspectBroadcast {
+    # fast push new published data
+
+    aspectHashDataList @0 :List(AspectHashDataList);
+    # the aspect data to publish / broadcast
+  }
+
+  struct MsgDirect {
+    # node-to-node message data
+
+    requestId @0 :Text;
+    # requestId for associating requests / responses
+
+    data @1 :Data;
+    # the content of the direct message
+  }
+
+  struct MsgQuery {
+    # dht query message data
+
+    requestId @0 :Text;
+    # requestId for associating requests / responses
+
+    entryAddress @1 :Data;
+    # the entryAddress being queried
+
+    data @2 :Data;
+    # the message content (either request or response)
+  }
+
+  # -- additional data types -- #
+
+  struct AspectHashList {
+    entryAddress @0 :Data;
+    # when referring to aspect hashes, we need them to be
+    # associated with an entryAddress
+
+    aspectHashList @1 :List(Data);
+    # the list of aspectHashes associated with the above entryAddress
+    # note this list is probably not comprehensive,
+    # it may be only those that fall within constraints, or only
+    # those that are being requested / responded with / etc.
+  }
+
+  struct AspectHashDataList {
+    entryAddress @0 :Data;
+    # when asking for aspect hash data, we also want to know the entryAddress
+
+    aspectHashDataList @1 :List(AspectHashData);
+    # the list of pairs of aspectHash / aspectData
+  }
+
+  struct AspectHashData {
```

Comment:
In the case of AspectHashData, storing the hash in the tuple probably means we need to calculate it as well to verify it, and if so why even store it?

Reply:
I agree we need to validate it if included.

It would simplify the api, and lower bandwidth to remove it.

Part of me likes the explicitness of having the hash in there, like, if a node had corrupted aspect data on its disk and published it, we could reject that at the protocol layer before even passing it off to core for validation. Buuut, if core is going to validate it anyway, perhaps we don't want to double that up?

Reply:
I'll go ahead and remove it for now, we can evaluate if that keeps making sense as we go forward.